sdk/core/System.ClientModel/src/ModelReaderWriter/ModelJsonConverter.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// A generic converter which allows <see cref="JsonSerializer"/> to be able to write and read any models that implement <see cref="IJsonModel{T}"/>.
+/// </summary>
+#if !NET5_0 // RequiresUnreferencedCode in net5.0 doesn't have AttributeTargets.Class as a target, but it was added in net6.0
+[RequiresUnreferencedCode("The constructors of the type being deserialized are dynamically accessed and may be trimmed.")]
+#endif
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+public class ModelJsonConverter : JsonConverter<IJsonModel<object>>
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+{
+    /// <summary>
+    /// Gets the <see cref="ModelReaderWriterOptions"/> used to read and write models.
+    /// </summary>
+    public ModelReaderWriterOptions Options { get; }
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="ModelJsonConverter"/> with a default options of <see cref="ModelReaderWriterOptions.Json"/>.
+    /// </summary>
+    public ModelJsonConverter()
+        : this(ModelReaderWriterOptions.Json) { }
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="ModelJsonConverter"/>.
+    /// </summary>
+    /// <param name="options">The <see cref="ModelReaderWriterOptions"/> to use.</param>
+    public ModelJsonConverter(ModelReaderWriterOptions options)
+    {
+        Options = options;
+    }
+
+    /// <inheritdoc/>
+    public override bool CanConvert(Type typeToConvert)
+    {
+        return !Attribute.IsDefined(typeToConvert, typeof(JsonConverterAttribute));
+    }
+
+    /// <inheritdoc/>
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+    public override IJsonModel<object> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+    {
+        using JsonDocument document = JsonDocument.ParseValue(ref reader);
```

Comment:
Is there a way to do it using Utf8JsonReader instead of JsonDocument?

Reply:
For deserialization not yet, this was punted from v1 but we have a prototype implementation here https://github.com/Azure/azure-sdk-for-net/pull/37970 that we can revive if we want to.

Only question is would we need that before this PR goes in?

The reasoning to punt on that initially was the deserialization was roughly the same perf as what we had before in our models since it also used JsonDocument and serialization was significantly faster.  We said we would add the utf8jsonreader integration later.

Reply:
I don't feel too strongly, just flagging since @KrzysztofCwalina recently flagged something similar in one of my PRs.  The BinaryData.FromString is possibly more of a perf issue, but [there is a forthcoming feature in .NET 9](https://github.com/dotnet/runtime/issues/77666) that IIRC may enable not transcoding to UTF16 -- maybe we move to that when available if we want to keep operating at the JsonDocument layer.